core/src/main/java/net/consensys/eventeum/chain/service/strategy/PubSubBlockSubscriptionStrategy.java

```diff
@@ -44,7 +44,7 @@ public Disposable subscribe() {
             final DefaultBlockParameter blockParam = DefaultBlockParameter.valueOf(latestBlock.get().getNumber());
 
             //New heads can only start from latest block so we need to obtain missing blocks first
-            blockSubscription = web3j.replayPastBlocksFlowable(blockParam, true)
+            blockSubscription = web3j.replayPastBlocksFlowable(blockParam, true).retry()
```

Comment:
I undetstand ur doubts due that I had the same. The problem, is that the underlying connection cannot be reconnected, only the logical connection. As example if on startup there is no connectivity, you cannot go further because no reconnect is available on web3j, only resubscribe which is mainly recreating.
As example today we had scenarios were with besu, no socket answer produced finally closed connection on websocke,  that could not be recovered, if no restarted,

This is why i think that retrying, makes some sense, and for connection  retry we need web3j to be  improved


io.reactivex.exceptions.UndeliverableException: The exception could not be delivered to the consumer because it has already canceled/disposed the flow or the exception has nowhere to go to begin with. Further re
ading: https://github.com/ReactiveX/RxJava/wiki/What's-different-in-2.0#error-handling | org.java_websocket.exceptions.WebsocketNotConnectedException
        at io.reactivex.plugins.RxJavaPlugins.onError(RxJavaPlugins.java:367)
        at io.reactivex.internal.disposables.CancellableDisposable.dispose(CancellableDisposable.java:51)
        at io.reactivex.internal.disposables.DisposableHelper.dispose(DisposableHelper.java:124)
        at io.reactivex.internal.disposables.SequentialDisposable.dispose(SequentialDisposable.java:72)
        at io.reactivex.internal.operators.flowable.FlowableCreate$BaseEmitter.cancel(FlowableCreate.java:301)
        at io.reactivex.internal.subscriptions.SubscriptionArbiter.drainLoop(SubscriptionArbiter.java:218)
        at io.reactivex.internal.subscriptions.SubscriptionArbiter.drain(SubscriptionArbiter.java:187)
        at io.reactivex.internal.subscriptions.SubscriptionArbiter.cancel(SubscriptionArbiter.java:179)
        at io.reactivex.internal.subscriptions.SubscriptionHelper.cancel(SubscriptionHelper.java:189)
        at io.reactivex.internal.subscribers.LambdaSubscriber.cancel(LambdaSubscriber.java:119)
        at io.reactivex.internal.subscribers.LambdaSubscriber.dispose(LambdaSubscriber.java:104)
        at net.consensys.eventeum.service.DefaultSubscriptionService.unsubscribeFilterSubscription(DefaultSubscriptionService.java:197)
        at net.consensys.eventeum.service.DefaultSubscriptionService.lambda$unsubscribeToAllSubscriptions$5(DefaultSubscriptionService.java:182)
        at java.util.concurrent.ConcurrentHashMap$ValuesView.forEach(ConcurrentHashMap.java:4707)
        at net.consensys.eventeum.service.DefaultSubscriptionService.unsubscribeToAllSubscriptions(DefaultSubscriptionService.java:180)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:343)
        at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:206)
        at com.sun.proxy.$Proxy135.unsubscribeToAllSubscriptions(Unknown Source)
        at net.consensys.eventeum.chain.service.health.NodeHealthCheckService.checkHealth(NodeHealthCheckService.java:113)
        at net.consensys.eventeum.chain.service.health.NodeHealthCheckService.lambda$new$0(NodeHealthCheckService.java:77)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: org.java_websocket.exceptions.WebsocketNotConnectedException
        at org.java_websocket.WebSocketImpl.send(WebSocketImpl.java:608)
        at org.java_websocket.WebSocketImpl.send(WebSocketImpl.java:585)
        at org.java_websocket.client.WebSocketClient.send(WebSocketClient.java:309)
        at org.web3j.protocol.websocket.WebSocketService.sendRequest(WebSocketService.java:172)
        at org.web3j.protocol.websocket.WebSocketService.sendAsync(WebSocketService.java:161)
        at org.web3j.protocol.websocket.WebSocketService.send(WebSocketService.java:140)
        at org.web3j.protocol.core.Request.send(Request.java:81)
        at org.web3j.protocol.core.filters.Filter.cancel(Filter.java:161)
        at io.reactivex.internal.disposables.CancellableDisposable.dispose(CancellableDisposable.java:48)

Reply:
Hmmm, if it can get into a state where the websocket can't be recovered, then theres probably a bug in the `NodeHealthCheckService` somewhere.  Do you have the logs from around the time that this occurred?  If so, could you email them to me?  Cheers!